Mathlib/RingTheory/MvPolynomial/Homogeneous.lean

```diff
@@ -532,13 +533,49 @@ theorem sum_homogeneousComponent :
   exact fun h => (coeff_eq_zero_of_totalDegree_lt h).symm
 #align mv_polynomial.sum_homogeneous_component MvPolynomial.sum_homogeneousComponent
 
-theorem homogeneousComponent_homogeneous_polynomial (m n : ℕ) (p : MvPolynomial σ R)
-    (h : p ∈ homogeneousSubmodule σ R n) : homogeneousComponent m p = if m = n then p else 0 := by
-  convert weightedHomogeneousComponent_weighted_homogeneous_polynomial m n p h
-#align mv_polynomial.homogeneous_component_homogeneous_polynomial MvPolynomial.homogeneousComponent_homogeneous_polynomial
+theorem homogeneousComponent_of_mem {m n : ℕ} {p : MvPolynomial σ R}
+    (h : p ∈ homogeneousSubmodule σ R n) :
+    homogeneousComponent m p = if m = n then p else 0 :=
+  weightedHomogeneousComponent_of_mem h
+#align mv_polynomial.homogeneous_component_homogeneous_polynomial MvPolynomial.homogeneousComponent_of_mem
 
 end HomogeneousComponent
 
 end
 
+noncomputable section GradedAlgebra
+
+/-- The homogeneous submodules form a graded ring.
+This instance is used by `DirectSum.commSemiring` and `DirectSum.algebra`. -/
+lemma HomogeneousSubmodule.gcomm_monoid :
+    SetLike.GradedMonoid (homogeneousSubmodule σ R) :=
+  WeightedHomogeneousSubmodule.gcomm_monoid
+
+variable [DecidableEq σ] [DecidableEq R]
+
+/-- The decomposition of `MvPolynomial σ R` into  homogeneous submodules -/
+abbrev decomposition  :
+    DirectSum.Decomposition (homogeneousSubmodule σ R) :=
+  weightedDecomposition R (1 : σ → ℕ)
+
+/-- Given a weight, `MvPolynomial σ R` as a graded algebra -/
+abbrev gradedAlgebra : GradedAlgebra (homogeneousSubmodule σ R) :=
+  weightedGradedAlgebra R (1 : σ → ℕ)
```

Comment:
This uses `weightedGradedAlgebra`, which has a `DecidableEq R` instance - is this avoidable at all?

Reply:
Good point, thank you. We have removed some instances `DecidableEq R` and `DecidableEq \sigma` which were only used in proofs.